proposals/inline-classes.md

```diff
@@ -82,6 +131,42 @@ Currently, inline classes must satisfy the following requirements:
 - Inline class cannot have inner classes
 - Inline class must be a toplevel class 
 
+Sidenotes:
+
+- Let's elaborate requirement to have public primary constructor and restriction of `init` blocks.
+For example, we want to have an inline class for some bounded value:
+    ```kotlin
+    inline class Positive(val value: Int) {
+        init { 
+            assert(value > 0) "Value isn't positive: $value" 
+        }
+    }
+  
+    fun foo(p: Positive) {}
+    ```
+    
+    Because of inlining, method `foo` have type `int` from Java POV, so we can pass to method `foo` everything we want and `init` 
+    block will not be executed. Since we cannot control behaviour of `init` block execution, we restrict it for inline classes.
+    
+    Unfortunately, it's not enough, because `init` blocks can be emulated via factory methods:
+    ```kotlin
+    inline class Positive private constructor(val value: Int) {
+        companion object {
+            fun create(x: Int) {
+                assert(x > 0) "Value isn't positive: x"
+                return Positive(x)  
+            }  
+        }
+    }
+  
+    fun foo(p: Positive) {}
+    ```
+    
+    Again, method `foo` have type `int` from Java POV, so we can indirectly create values of type `Positive` 
+    even with the presence of private constructor.
+    
+    To make behaviour more predictable and consistent with Java, we demand public primary constructor and restrict `init` blocks.
```

Comment:
Why not allow init blocks and private constructors by enforcing boxed type when used from Java? This would work similarly to inline functions that are inline when invoked in Kotlin, but not when called from Java or the debugger.

Reply:
Yes, this approach is nice, some kind of "closed world" for inline classes. 
But note that inline functions cannot be virtual, while inline class types can be used everywhere. To enforce boxing for Java we'll have to duplicate methods, plus there are some issues with arrays (if we'll allow them, we'll have to box each element of an array on the border)

I'd say that we should think more about this approach, at least to understand our design space

Reply:
@zarechenskiy It should not duplicate methods, but rather, add a method that takes the boxed type that is visible in Java, which unboxes the value and call the static method that takes the unboxed value.
In case the defined method is also `inline`, the method with the boxed type would contain all the logic directly, since no method for the unboxed type would be generated (since they would all be inlined at call places).